File: src/leetcode/java/BinaryTreePreorderTraversal_144.java
package leetcode.java;

import java.util.LinkedList;
import java.util.List;

public class BinaryTreePreorderTraversal_144 {
  public List<Integer> preorderTraversal(TreeNode root) {
    List<Integer> list = new LinkedList<Integer>();
    addToList(root, list);
    return list;
  }

  public static void addToList(TreeNode root, List<Integer> list) {
    if (root == null) return;
    if (root.val != '#') list.add(root.val);
    addToList(root.left, list);
    addToList(root.right, list);
  }
}

File: src/leetcode/python/BinaryTreePreorderTraversal_144.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinaryTreePreorderTraversal_144(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def addNode(node):
            if node is None:
                return
            if node.val != '#':
                res.append(node.val)
            addNode(node.left)
            addNode(node.right)
        res = []
        addNode(root)
        return res
